### PR TITLE
Integrate coverage configuration and tests in CI pipeline

### PR DESCRIPTION
**Pull request by OlenaYefymenko**
This patch provides the coverage configuration, updates the project dependencies accordingly, and adds tests to the CI pipeline. This PR sets the stage for the next, where test functions will be integrated into the project.

Related to #18

### CODE REVIEW

**Comment by OlenaYefymenko** [comment]
@webknjaz I've implemented the `coverage` configuration as we did in my previous project. However, you mentioned an option regarding this:

```
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys


def main():
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "myproj.settings")

    # MyProject Customization: run coverage.py around tests automatically
    try:
        command = sys.argv[1]
    except IndexError:
        command = "help"

    running_tests = command == "test"
    if running_tests:
        from coverage import Coverage

        cov = Coverage()
        cov.erase()
        cov.start()

    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)

    if running_tests:
        cov.stop()
        cov = Coverage()
        cov.save()
        covered = cov.report()
        if covered < 100:
            raise SystemExit(1)


if __name__ == "__main__":
    main()
```

I have doubts about whether I should keep the configuration as it is in the current PR or add it into `manage.py` using `contextlib.nullcontext()`. However, I've read that using a `coveragerc` file is preferable.

**Comment by webknjaz** [comment]
Use the coverage file for the configuration but integrate nullcontext for enabling coverage collection conditionally. Also, use the contextmanager decorator to move that to a separate helper function.

**Comment by OlenaYefymenko** [comment]
@webknjaz I've just added what we wrote in the live coding session.

File: .coveragerc [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+skip_covered = false
+
+[paths]
+source =
+  mysite
+
+[report]
+skip_covered = true
+show_missing = true
+exclude_lines =
```
Comment (webknjaz) [inline .coveragerc:12]: There's a new setting in coveragepy called `exclude_also`. There are default entries that coveragepy sets for `exclude_lines`. If you supply your own list, it replaces those entries, but with `exclude_also`, it extends the default entries with yours. It's considered better now. Also, in the future, it might be nice to integrate https://pypi.org/p/covdefaults which provides many popular settings automatically.
But for now, this small change should suffice:
```suggestion
exclude_also =
```

File: .coveragerc [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  if __name__ == ['"]__main__['"]:$
+
+[run]
+branch = true
+cover_pylib = false
+dynamic_context = test_function
+parallel = true
+relative_files = true
+source =
+  tests
```
Comment (webknjaz) [inline .coveragerc:27]: You don't have this top-level folder so it's probably unnecessary to use this setting. OTOH, it may be nice to move the tests to the repository root so perhaps keeping it as it is is warranted.

File: .coveragerc [hunk 3]
```diff
@@ -0,0 +1,29 @@
+[html]
+show_contexts = true
+skip_covered = false
+
+[paths]
+source =
+  mysite
```
Comment (webknjaz) [inline .coveragerc:7]: This section is not needed. It's used to map files in your repository with their site-packages copies. But since a Django project isn't `pip install`ed, there should be no need to configure this mapping.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -6,6 +6,24 @@ on:  # yamllint disable-line rule:truthy
...
+    - name: Fetch the src
+      uses: actions/checkout@v3
+    - name: Set up Python 3.11
+      uses: actions/setup-python@v4
+      with:
+        python-version: 3.11
+
+    - name: Install requirements
+      run: python -m pip install -r requirements/base.in
+           -c requirements/base.txt
```
Comment (webknjaz) [inline .github/workflows/ci.yml:22]: Let's make the formatting more intuitive here:
```suggestion
      run: >-
        python -m pip install
        -r requirements/base.in
        -c requirements/base.txt
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -17,6 +35,10 @@ jobs:
       with:
         python-version: 3.11
 
+    - name: Install requirements
+      run: python -m pip install -r requirements/base.in
+           -c requirements/base.txt
```
Comment (webknjaz) [inline .github/workflows/ci.yml:40]:
```suggestion
      run: >-
        python -m pip install
        -r requirements/base.in
        -c requirements/base.txt
```

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -47,6 +47,7 @@ repos:
     - --django-settings-module=mysite.settings
     additional_dependencies:
     - pylint_django[with_django] >= 2.5.3
+    - coverage
```
Comment (webknjaz) [inline .pre-commit-config.yaml:50]: It's usually nicer to have these entries sorted:
```suggestion
    - coverage
    - pylint_django[with_django] >= 2.5.3
```

File: manage.py [hunk 7]
```diff
@@ -1,9 +1,45 @@
-#!/usr/bin/env python
```
Comment (webknjaz) [inline manage.py:1]: Don't remove this. It's unrelated to coverage and is necessary for running the script as `./manage.py`.

File: manage.py [hunk 8]
```diff
@@ -1,9 +1,45 @@
...
 """Django's command-line utility for administrative tasks."""
-import os
 import sys
+import os
```
Comment (webknjaz) [inline manage.py:3]: Keep the entries sorted. Also, making formatting changes in the same PR as functional changes is considered a very bad tone because this makes it _much harder_ to compare the versions during review.
```suggestion
import os
import sys
```

File: requirements/base.in [hunk 9]
```diff
@@ -1 +1,2 @@
 Django ~= 3.0
+coverage==7.3.2
```
Comment (webknjaz) [inline requirements/base.in:2]: Let's not add test-only dependencies to the base requirements. Use a separate requirements file for non-production envs. Also, no need to pin the version in this file because it's an input file and the pins need to be updateable by pip-tools in the future, don't prevent this.

File: requirements/base.txt
Comment (webknjaz) [inline requirements/base.txt:1]: Per above, this lockfile shouldn't be updated.

File: manage.py [hunk 10]
```diff
@@ -1,9 +1,45 @@
...
+    """
+    # pylint: disable-next=import-outside-toplevel
+    from coverage import Coverage
+
+    cov = Coverage()
+    cov.erase()
+    with cov.collect():
+        yield
+    cov.save()
+    print(f'Coverage is {cov.report()}%')  # xml, html
```
Comment (webknjaz) [inline manage.py:22]: The comment is pointless here, because it doesn't mean anything, nor is related to printing out the text on this line.
```suggestion
    print(f'Coverage is {cov.report()}%')
```

File: manage.py [hunk 11]
```diff
@@ -1,9 +1,45 @@
...
+
+@contextmanager
+def coverage_context():
+    """Context manager to collect code coverage data.
+
+    Uses coverage package to collect coverage, generate reports
+    and print coverage percentage.
+    """
+    # pylint: disable-next=import-outside-toplevel
+    from coverage import Coverage
```
Comment (webknjaz) [inline manage.py:15]: It's possible to update `.pylintrc` instead of making the inline pylint-disable comment:
```ini
allow-any-import-level =
  coverage,
```
But it's up to you, keeping it here is also fine.

File: manage.py [hunk 12]
```diff
@@ -1,9 +1,45 @@
-#!/usr/bin/env python
 """Django's command-line utility for administrative tasks."""
-import os
 import sys
+import os
+from contextlib import nullcontext, contextmanager
```
Comment (webknjaz) [inline manage.py:4]: An alternative solution to importing coverage is
```suggestion
from contextlib import contextmanager, nullcontext, suppress

with suppress(ImportError):
    from coverage import Coverage
```
This way, it wouldn't be imported in the function context, which may be perceived as a bit more elegant.
Though, the downside is that the import would be attempted even when unneeded (for non-test subcommands).

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: .coveragerc (hunk 2 above)
Comment (OlenaYefymenko) [reply .coveragerc:27]: Okay, I will create the directory in the next PR.

Comment (webknjaz) [reply .coveragerc:27]: Marking this as resolved.

File: manage.py (hunk 10 above)
Comment (OlenaYefymenko) [reply manage.py:22]: Agree

File: manage.py (hunk 11 above)
Comment (OlenaYefymenko) [reply manage.py:15]: Ok, I will change it. Thanks.

File: requirements/base.in (hunk 9 above)
Comment (OlenaYefymenko) [reply requirements/base.in:2]: ок

**Comment by OlenaYefymenko** [comment]
Sorry, in next PR I will add  `# yamllint disable-line rule:truthy`.

**Comment by webknjaz** [comment]
I got some notification regarding a "truthy" yamllint rule and can't find that comment to see the context here. What was that about? Can you link it?

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -6,6 +6,32 @@ on:  # yamllint disable-line rule:truthy
...
+    - name: Set up Python 3.11
+      uses: actions/setup-python@v4
+      with:
+        python-version: 3.11
+
+    - name: Install requirements
+      run: >-
+        python -m pip install
+        -r requirements/base.in
+        -c requirements/base.txt
```
Comment (webknjaz) [inline .github/workflows/ci.yml:24]: These shouldn't be separate, otherwise, they'll be overwritten by the following `pip install` with possibly different versions of things.

File: .github/workflows/ci.yml [hunk 14]
```diff
@@ -17,6 +43,18 @@ jobs:
       with:
         python-version: 3.11
 
+    - name: Install requirements
+      run: >-
+        python -m pip install
+        -r requirements/base.in
+        -c requirements/base.txt
```
Comment (webknjaz) [inline .github/workflows/ci.yml:50]: Same here: these versions must be consistent with the other constraints file.

File: manage.py (hunk 7 above)
Comment (webknjaz) [reply manage.py:1]: Marking this as resolved.

File: requirements/base.in [hunk 15]
```diff
@@ -1 +1 @@
-Django ~= 3.0
```
Comment (webknjaz) [inline requirements/base.in:1]: How is this change related to the PR?

File: requirements/base.txt [hunk 16]
```diff
@@ -2,12 +2,12 @@
 # This file is autogenerated by pip-compile with Python 3.11
 # by the following command:
 #
-#    pip-compile base.in
+#    pip-compile requirements/base.in
```
Comment (webknjaz) [inline requirements/base.txt:5]: Looks like you're running `pip-tools` from a different directory, then before. Be consistent so that it doesn't generate pointless changes.

File: requirements/dev.in
Comment (webknjaz) [inline requirements/dev.in:1]: Don't forget to include the base requirements so that the pins are consistent and don't conflict with each other.

**Comment by OlenaYefymenko** [comment]
> I got some notification regarding a "truthy" yamllint rule and can't find that comment to see the context here. What was that about? Can you link it?

Yes, first I pushed without this comment and CI linters crashed, but then I fixed this and deleted the comment

**Comment by webknjaz** [comment]
It's best to communicate what's happening. Deleting things only complicates everything.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (OlenaYefymenko) [reply .github/workflows/ci.yml:24]: > These shouldn't be separate, otherwise, they'll be overwritten by the following `pip install` with possibly different versions of things.

I'm sorry, but I don't clearly understand how the installation of requirements should look in the configuration file. Here is your previous fix https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425681784. 
I'll try to find some examples and docs, maybe.

File: requirements/dev.in
Comment (OlenaYefymenko) [reply requirements/dev.in:1]: Ok

File: .github/workflows/ci.yml (hunk 13 above)
Comment (OlenaYefymenko) [reply .github/workflows/ci.yml:24]: `run: python -m pip install -r requirements/base.in -c requirements/base.txt`
like this variant

Comment (webknjaz) [reply .github/workflows/ci.yml:24]: You mustn't install base explicitly/directly.

File: requirements/dev.in
Comment (webknjaz) [reply requirements/dev.in:1]: Add the input file, not the constraints one.

File: requirements/base.in (hunk 15 above)
Comment (OlenaYefymenko) [reply requirements/base.in:1]: I remember about atomicity, but in this PR I added `coverage` and restructured requirements. It may be complicated to separate them.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (OlenaYefymenko) [reply .github/workflows/ci.yml:24]: Yes, I understood. Thank you)

File: requirements/base.in (hunk 15 above)
Comment (webknjaz) [reply requirements/base.in:1]: This file just doesn't need any changes at all. No need to separate anything from anywhere. You add other deps into a different lockfile.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (webknjaz) [reply .github/workflows/ci.yml:24]: > Here is your previous fix [#31 (comment)](https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425681784).

That was just an example of a formatting change, orthogonal to the fact that the step shouldn't be included, actually.

File: requirements/dev.in
Comment (webknjaz) [reply requirements/dev.in:1]: Marking this as resolved.

**Review by webknjaz — CHANGES_REQUESTED** [review]
Reminding to undo reshuffling the imports and address other comments.

**Comment by OlenaYefymenko** [comment]
> Reminding to undo reshuffling the imports and address other comments.

I've merged the commits with the import into the previous, resolved conflicts. I hope this is appropriate for this issue. Also, I assume I've checked all comments, except for this one: https://github.com/kpi-web-guild/django-girls-blog-OlenaEfymenko/pull/31#discussion_r1425699563
Can we postpone addressing that one?

File: .pre-commit-config.yaml [hunk 17]
```diff
@@ -46,8 +46,10 @@ repos:
     args:
     - --django-settings-module=mysite.settings
     additional_dependencies:
+    - coverage
     - pylint_django[with_django] >= 2.5.3
 
+
```
Comment (webknjaz) [inline .pre-commit-config.yaml:52]: Here.

File: manage.py [hunk 18]
```diff
@@ -1,9 +1,47 @@
 #!/usr/bin/env python
+
```
Comment (webknjaz) [inline manage.py:2]: Here.

File: requirements/base.txt
Comment (webknjaz) [reply requirements/base.txt:1]: Here.

**Review by webknjaz — CHANGES_REQUESTED** [review]
These are some unrelated changes I spotted.

File: requirements/base.txt
Comment (OlenaYefymenko) [reply requirements/base.txt:1]: > Per above, this lockfile shouldn't be updated.

Ok, but in this PR, I should separate the `coverage` into another section of the dep. That's why the `base.txt` file also needs updating after this. Maybe we can keep it this way?

File: .pre-commit-config.yaml (hunk 17 above)
Comment (OlenaYefymenko) [reply .pre-commit-config.yaml:52]: I assume that we are integrating `coverage` into the project within the current PR. Additionally, in this commit, we should modify the `pre-commit tool` by adding `coverage`. Sorry, I'm a bit confused.

File: manage.py (hunk 18 above)
Comment (OlenaYefymenko) [reply manage.py:2]: Thank you. I've squashed commits.

File: requirements/base.txt
Comment (webknjaz) [reply requirements/base.txt:1]: No, it shouldn't be updated. You already reverted most of the changes except for comment lines with file paths.

File: .pre-commit-config.yaml (hunk 17 above)
Comment (webknjaz) [reply .pre-commit-config.yaml:52]: Exactly. Not formatting changes in disconnected places.

File: manage.py (hunk 18 above)
Comment (webknjaz) [reply manage.py:2]: Okay, now remove the formatting changes.

File: requirements/base.txt (hunk 16 above)
Comment (webknjaz) [reply requirements/base.txt:5]: Adding a 👍 doesn't count as reverting an unrelated change.

Comment (OlenaYefymenko) [reply requirements/base.txt:5]: > Adding a 👍 doesn't count as reverting an unrelated change.

Yes, it looks like this pull request needs a lot of action with fixing commits.

File: manage.py (hunk 18 above)
Comment (OlenaYefymenko) [reply manage.py:2]: I'm sorry, yes, I assume when I resolved conflicts I didn't finish this issue with `#!/usr/bin/env python`.
Also, I didn't notice just the empty line in the config. At first, I guess you meant about `coverage`. Now I hope the PR is atomic. I understand that `pylint` was added in the dependencies, but it's just one minor change not closely related to this PR.

**Review by webknjaz — APPROVED** [review]

File: .pre-commit-config.yaml (hunk 17 above)
Comment (webknjaz) [reply .pre-commit-config.yaml]: Marking this as resolved.

File: manage.py (hunk 18 above)
Comment (webknjaz) [reply manage.py]: Marking this as resolved.

File: .coveragerc (hunk 3 above)
Comment (webknjaz) [reply .coveragerc]: Marking this as resolved.

File: manage.py (hunk 8 above)
Comment (webknjaz) [reply manage.py]: Marking this as resolved.

File: requirements/base.in (hunk 9 above)
Comment (webknjaz) [reply requirements/base.in]: Marking this as resolved.

File: requirements/base.txt
Comment (webknjaz) [reply requirements/base.txt:1]: Marking this as resolved.

File: manage.py (hunk 11 above)
Comment (webknjaz) [reply manage.py]: Marking this as resolved.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (webknjaz) [reply .github/workflows/ci.yml]: Marking this as resolved.

File: requirements/base.in (hunk 15 above)
Comment (webknjaz) [reply requirements/base.in]: Marking this as resolved.

File: .github/workflows/ci.yml (hunk 14 above)
Comment (webknjaz) [reply .github/workflows/ci.yml]: Marking this as resolved.

File: requirements/base.txt (hunk 16 above)
Comment (webknjaz) [reply requirements/base.txt]: Marking this as resolved.
